backend/generalhw.pm

```diff
@@ -27,39 +28,43 @@ sub get_cmd ($self, $cmd) {
     my $dir = get_required_var('GENERAL_HW_CMD_DIR');
     die 'GENERAL_HW_CMD_DIR is not pointing to a directory' unless -d $dir;
 
-    my %GENERAL_HW_ARG_VARIABLES_BY_CMD = ('GENERAL_HW_FLASH_CMD' => 'GENERAL_HW_FLASH_ARGS', 'GENERAL_HW_SOL_CMD' => 'GENERAL_HW_SOL_ARGS', 'GENERAL_HW_POWERON_CMD' => 'GENERAL_HW_POWERON_ARGS', 'GENERAL_HW_POWEROFF_CMD' => 'GENERAL_HW_POWEROFF_ARGS');
+    my %GENERAL_HW_ARG_VARIABLES_BY_CMD = (
+        'GENERAL_HW_FLASH_CMD' => 'GENERAL_HW_FLASH_ARGS',
+        'GENERAL_HW_SOL_CMD' => 'GENERAL_HW_SOL_ARGS',
+        'GENERAL_HW_POWERON_CMD' => 'GENERAL_HW_POWERON_ARGS',
+        'GENERAL_HW_POWEROFF_CMD' => 'GENERAL_HW_POWEROFF_ARGS',
+        'GENERAL_HW_STORE_ASSET_CMD' => 'GENERAL_HW_STORE_ASSET_ARGS',
+    );
     my $args = get_var($GENERAL_HW_ARG_VARIABLES_BY_CMD{$cmd}) if get_var($GENERAL_HW_ARG_VARIABLES_BY_CMD{$cmd});
 
-    # Append HDD infos to flash script
-    if ($cmd eq 'GENERAL_HW_FLASH_CMD' and get_var('HDD_1')) {
-        my $numdisks = get_var('NUMDISKS') // 1;
-        for my $i (1 .. $numdisks) {
-            # Pass path of HDD
-            $args .= " " . get_required_var("HDD_$i");
-            # Pass size of HDD
-            my $size = get_var("HDDSIZEGB_$i");
-            $size //= get_var('HDDSIZEGB') // 10;
-            $args .= " $size" . 'G';
-        }
-    }
-
     $cmd = get_required_var($cmd);
     $cmd = "$dir/" . basename($cmd);
     $cmd .= " $args" if $args;
     return $cmd;
 }
 
-sub run_cmd ($self, @args) {
-    my @full_cmd = split / /, $self->get_cmd($args[0]);
+sub run_cmd ($self, $cmd, @extra_args) {
+    my @full_cmd = split / /, $self->get_cmd($cmd);
 
-    my ($stdin, $stdout, $stderr, $ret);
-    eval { $ret = IPC::Run::run([@full_cmd], \$stdin, \$stdout, \$stderr) };
-    die "Unable to run command '@full_cmd' (deduced from test variable $args[0]): $@\n" if $@;
+    push @full_cmd, @extra_args;
+
+    my ($stdout, $stderr, $ret);
+    eval {
+        my ($path, @args) = @full_cmd;
+        my $p = process(execute => $path, args => \@args)->start;
+        $p->to_ioloop->on(read => sub {
+                $stdout .= $p->read_all_stdout;
+                $stderr .= $p->read_all_stderr;
+        });
+        Mojo::IOLoop->singleton->start() unless Mojo::IOLoop->singleton->is_running;
```

Comment:
`start` will block until `stop` is called which doesn't seem to be the case here. So this code will likely block here infinitely (unless the loop is already running anyways). I would have to check the Mojo::IOLoop::ReadWriteProcess documentation (and possibly implementation details) myself to tell you what the right approach would be, tough.

Reply:
> I would have to check the Mojo::IOLoop::ReadWriteProcess documentation (and possibly implementation details) myself to tell you what the right approach would be, tough.

I did. The documentation is not much helping (just one sentence for most methods, and no example how to "run a program and collect all its output" - just some examples for interactive things, like read one line, write one line etc). Inspecting implementation got me to this version, since `read_all_stdout` isn't really reading all stdout - it reads all stdout that the process produced so far (and exit early if nothing new is available).

I'm mostly worried about the possible deadlock case, where the process produces a lot of output on stdout _and_ stderr, filling either of the pipes, and the other end is waiting reading from the other pipe...

> So this code will likely block here infinitely (unless the loop is already running anyways).

Can I assume (and assert for, maybe) that the loop is already running? It isn't clear to me from code inspection... But given this version works, I guess it is the case.

Reply:
The documentation (and implementation) both are not very helpful, indeed.

> Can I assume (and assert for, maybe) that the loop is already running? It isn't clear to me from code inspection... But given this version works, I guess it is the case.

I don't think the Mojo IO loop is running in the backend process (which in fact runs its own "capture loop" using `IO::Select`). You could try to read from the `IO::Handle` (`$process->read_stream`) directly. Or you stop the `Mojo::IOLoop` in some event handler (e.g. the process' `stop` event).

Reply:
> Otherwise sometimes Mojo's loop collects the exit code of a process started by IPC::Run::run, which results in a false failure.

Note that it is not Mojo's loop which does that. It is the handler we register for the `Mojo::IOLoop::ReadWriteProcess::Session` which uses Mojo's event system but not the event loop (at least that's how it looks like after a brief look; the event is emitted from the `SIG_CHLD` handler).

I'm wondering whether switching to RWP makes any difference in that regard.

I'm also wondering whether the failure is really caused by the `SIG_CHLD` handler. All we do in that handler is logging.

Reply:
What I really want to achieve here, is to avoid `Mojo` collecting exit code _elsewhere_ of the for child process started here. What happens is that `session->on( collected_orphan => ...)` in `backend/driver.pm` is collecting it. If `Mojo::IOLoop` isn't normally running there, maybe I could prevent that thing from happening either? I guess it's registering SIGCHLD handler - maybe I could temporarily disable it (`$SIG[CHLD] = "DEFAULT"`?) instead of switching to `Mojo::IOLoop::ReadWriteProcess`? That would still risk some weird interactions (`IPC::Run` collecting an exit code that was supposed to be collected by `Mojo::IOLoop`), but maybe that's less of an issue?

What would you prefer:
1. temporarily starting (and stopping) `Mojo::IOLoop`, or
2. trying to make it work with `IPC:Run`

?

Reply:
> All we do in that handler is logging.

Since the message does include the PID and exit status, the exit status was retrieved, so `IPC::Run` can't do it anymore.

Reply:
> I'm wondering whether switching to RWP makes any difference in that regard.

I guess it does make a difference, because RWP can coordinate with the signal handler registered by another part of `Mojo::IOLoop`.

Reply:
Then it'll likely be the easiest to start/stop the event loop as it would indeed avoid unwanted interactions between the process management libraries.

Reply:
You can always temporarily override the signal handler in the `IPC::Run` block, it's blocking after all. Just `local $SIG{CHLD} = ...` or similar at the start of it.